src/calendar-app.ts

```diff
@@ -0,0 +1,774 @@
+): RouteResolver {
+	requireLogin = requireLogin ?? true
+	let cache: RouteCache | null
+
+	// a bit of context for why we do things the way we do. Constraints:
+	//  - view must be imported async in onmatch
+	//  - view shall not be created manually, we do not want to hold on to the instance
+	//  - we want to pass additional parameters to the view
+	//  - view should not be created twice and neither its dependencies
+	//  - we either need to call updateUrl or pass requestedPath and args as attributes
+	return {
+		// onmatch() is called for every URL change
+		async onmatch(args: Record<string, Dict>, requestedPath: string): Promise<Class<ComponentType> | null> {
+			// enforce valid login state first.
+			// we have views with requireLogin: true and views with requireLogin: false, each of which enforce being logged in or being logged out respectively.
+			// in the logout case (where requireLogin: false) this will force a reload.
+			// the login view is special in that it has requirelogin: false, but can be logged in after account creation during signup.
+			// to handle back button presses where the user decides to stay on the page after all (we show a confirmation)
+			// we need to prevent the logout/reload. this is the purpose of the keepSession argument.
+			// the signup wizard that sets it handles the session itself.
+			if (requireLogin && !logins.isUserLoggedIn()) {
+				forceLogin(args, requestedPath)
+				return null
+			} else if (!requireLogin && logins.isUserLoggedIn() && !args.keepSession) {
+				await disableErrorHandlingDuringLogout()
+				await logins.logout(false)
+				windowFacade.reload(args)
+				return null
+			} else {
+				const prepared = await prepareRoute(cache)
+				cache = prepared.cache
+				return prepared.component
+			}
+		},
+		// render() is called on every render
+		render(vnode: Vnode<ComponentType>): Children {
+			const args = m.route.param()
+			const requestedPath = m.route.get()
+			// result of onmatch() is passed into m() by mthril and then given to us here
+			// It is not what we want as we want to pass few things to it but it's harmless because
+			// it just creates a vnode but doesn't render it.
+			// What we do is grab the class from that vnode. We could have done it differently but this
+			// way we don't do any more caching than Mithril would do anyway.
+
+			// TS can't prove that it's the right component and the mithril typings are generally slightly broken
+			const c = vnode.tag as unknown as Class<ClassComponent<FullAttrs>>
+
+			// downcast because we ts can't really prove or enforce that additional attrs have compatible requestedPath and args
+			const attrs = { requestedPath, args, ...prepareAttrs(assertNotNull(cache)) } as FullAttrs
+			return m(
+				root,
+				m(c, {
+					...attrs,
+					oncreate({ state }: VnodeDOM<FullAttrs, ComponentType>) {
+						window.tutao.currentView = state
+					},
+				}),
+			)
+		},
+	}
+}
+
+function makeOldViewResolver(
```

Comment:
this is only used with (non-mobile) webauthn

Reply:
While the focus is the apps right now, we are still figuring out if/how we will split desktop and web later so I want to leave this type of stuff in, though I will make a comment on it.

Reply:
okay thank you, then please also disregard my other comment about paths